debian/bongo.1
.\" Man page generated from reStructuredText.
.
.TH "BONGO" "1" "January 30, 2015" "3.0" "bongodb-manual"
.SH NAME
bongo \- BongoDB Shell
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH DESCRIPTION
.sp
\fBbongo\fP is an interactive JavaScript shell interface to
BongoDB, which provides a powerful interface for systems
administrators as well as a way for developers to test queries and
operations directly with the database. \fBbongo\fP also provides
a fully functional JavaScript environment for use with a BongoDB. This
document addresses the basic invocation of the \fBbongo\fP shell
and an overview of its usage.
.SH OPTIONS
.SS Core Options
.INDENT 0.0
.TP
.B bongo
.UNINDENT
.INDENT 0.0
.TP
.B \-\-shell
Enables the shell interface. If you invoke the \fBbongo\fP command
and specify a JavaScript file as an argument, or use \fI\%\-\-eval\fP to
specify JavaScript on the command line, the \fI\%\-\-shell\fP option
provides the user with a shell prompt after the file finishes executing.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-nodb
Prevents the shell from connecting to any database instances. Later, to
connect to a database within the shell, see
\fIbongo\-shell\-new\-connections\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-norc
Prevents the shell from sourcing and evaluating \fB~/.bongorc.js\fP on
start up.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-quiet
Silences output from the shell during the connection process.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-port <port>
Specifies the port where the \fBbongod\fP or \fBbongos\fP
instance is listening. If \fI\-\-port\fP is not specified,
\fBbongo\fP attempts to connect to port \fB27017\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-host <hostname>
Specifies the name of the host machine where the \fBbongod\fP or
\fBbongos\fP is running. If this is not specified,
\fBbongo\fP attempts to connect to a BongoDB process running on
the localhost.
.sp
To connect to a replica set, specify the \fBreplica set name\fP and a seed list of set members. Use the
following form:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
<replSetName>/<hostname1><:port>,<hostname2><:port>,<...>
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-eval <javascript>
Evaluates a JavaScript expression that is specified as an argument.
\fBbongo\fP does not load its own environment when evaluating code.
As a result many options of the shell environment are not available.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-username <username>, \-u <username>
Specifies a username with which to authenticate to a BongoDB database
that uses authentication. Use in conjunction with the \fB\-\-password\fP and
\fB\-\-authenticationDatabase\fP options.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-password <password>, \-p <password>
Specifies a password with which to authenticate to a BongoDB database
that uses authentication. Use in conjunction with the \fI\-\-username\fP
and \fI\-\-authenticationDatabase\fP options. To force \fBbongo\fP to
prompt for a password, enter the \fI\-\-password\fP option as the
last option and leave out the argument.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-help, \-h
Returns information on the options and use of \fBbongo\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-version
Returns the \fBbongo\fP release number.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-verbose
Increases the verbosity of the output of the shell during the connection
process.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ipv6
Enables IPv6 support and allows the \fBbongo\fP to connect to the
BongoDB instance using an IPv6 network. All BongoDB programs and
processes disable IPv6 support by default.
.UNINDENT
.INDENT 0.0
.TP
.B <db address>
Specifies the "database address" of the database to connect to. For
example:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
bongo admin
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The above command will connect the \fBbongo\fP shell to the
\fIadmin database\fP on the local machine. You may specify a remote
database instance, with the resolvable hostname or IP address. Separate
the database name from the hostname using a \fB/\fP character. See the
following examples:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
bongo bongodb1.example.net
bongo bongodb1/admin
bongo 10.8.8.10/test
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This syntax is the \fIonly\fP way to connect to a specific database.
.sp
To specify alternate hosts and a database, you must use this syntax and cannot
use \fI\-\-host\fP or \fI\-\-port\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B <file.js>
Specifies a JavaScript file to run and then exit. Generally this should
be the last option specified.
.INDENT 7.0
.INDENT 3.5
.SS Optional
.sp
To specify a JavaScript file to execute \fIand\fP allow
\fBbongo\fP to prompt you for a password using
\fI\-\-password\fP, pass the filename as the first parameter with
\fI\-\-username\fP and \fI\-\-password\fP as the last options, as
in the following:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
bongo file.js \-\-username username \-\-password
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
Use the \fI\%\-\-shell\fP option to return to a shell after the file
finishes running.
.UNINDENT
.SS Authentication Options
.INDENT 0.0
.TP
.B \-\-authenticationDatabase <dbname>
New in version 2.4.

.sp
Specifies the database that holds the user\(aqs credentials.
.sp
If you do not specify a value for \fI\-\-authenticationDatabase\fP, \fBbongo\fP uses the database
specified in the connection string.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-authenticationMechanism <name>
\fIDefault\fP: BONGODB\-CR
.sp
New in version 2.4.

.sp
Changed in version 2.6: Added support for the \fBPLAIN\fP and \fBBONGODB\-X509\fP authentication
mechanisms.

.sp
Specifies the authentication mechanism the \fBbongo\fP instance uses to
authenticate to the \fBbongod\fP or \fBbongos\fP\&.
.TS
center;
|l|l|.
_
T{
Value
T}	T{
Description
T}
_
T{
BONGODB\-CR
T}	T{
BongoDB challenge/response authentication.
T}
_
T{
BONGODB\-X509
T}	T{
BongoDB SSL certificate authentication.
T}
_
T{
PLAIN
T}	T{
External authentication using LDAP. You can also use \fBPLAIN\fP
for authenticating in\-database users. \fBPLAIN\fP transmits
passwords in plain text. This mechanism is available only in
\fI\%BongoDB Enterprise\fP\&.
T}
_
T{
GSSAPI
T}	T{
External authentication using Kerberos. This mechanism is
available only in \fI\%BongoDB Enterprise\fP\&.
T}
_
.TE
.UNINDENT
.INDENT 0.0
.TP
.B \-\-gssapiHostName
New in version 2.6.

.sp
Specify the hostname of a service using \fBGSSAPI/Kerberos\fP\&. \fIOnly\fP required if the hostname of a machine does
not match the hostname resolved by DNS.
.sp
This option is available only in BongoDB Enterprise.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-gssapiServiceName
New in version 2.6.

.sp
Specify the name of the service using \fBGSSAPI/Kerberos\fP\&. Only required if the service does not use the
default name of \fBbongodb\fP\&.
.sp
This option is available only in BongoDB Enterprise.
.UNINDENT
.SS SSL Options
.INDENT 0.0
.TP
.B \-\-ssl
New in version 2.2.

.sp
Enables connection to a \fBbongod\fP or \fBbongos\fP that has
SSL support enabled.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyFile <filename>
New in version 2.4.

.sp
Specifies the \fB\&.pem\fP file that contains both the SSL certificate
and key. Specify the file name of the \fB\&.pem\fP file using relative
or absolute paths.
.sp
This option is required when using the \fB\-\-ssl\fP option to connect
to a \fBbongod\fP or \fBbongos\fP that has
\fBCAFile\fP enabled \fIwithout\fP
\fBallowConnectionsWithoutCertificates\fP\&.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyPassword <value>
New in version 2.4.

.sp
Specifies the password to de\-crypt the certificate\-key file (i.e.
\fB\-\-sslPEMKeyFile\fP). Use the \fI\-\-sslPEMKeyPassword\fP option only if the
certificate\-key file is encrypted. In all cases, the \fBbongo\fP will
redact the password from all logging and reporting output.
.sp
Changed in version 2.6: If the private key in the PEM file is encrypted and you do not
specify the \fI\-\-sslPEMKeyPassword\fP option, the \fBbongo\fP will prompt for a
passphrase. See \fIssl\-certificate\-password\fP\&.

.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCAFile <filename>
New in version 2.4.

.sp
Specifies the \fB\&.pem\fP file that contains the root certificate chain
from the Certificate Authority. Specify the file name of the
\fB\&.pem\fP file using relative or absolute paths.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.sp
\fBWARNING:\fP
.INDENT 7.0
.INDENT 3.5
If the \fBbongo\fP shell or any other tool that connects to
\fBbongos\fP or \fBbongod\fP is run without
\fI\-\-sslCAFile\fP, it will not attempt to validate
server certificates. This results in vulnerability to expired
\fBbongod\fP and \fBbongos\fP certificates as well as to foreign
processes posing as valid \fBbongod\fP or \fBbongos\fP
instances. Ensure that you \fIalways\fP specify the CA file against which
server certificates should be validated in cases where intrusion is a
possibility.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCRLFile <filename>
New in version 2.4.

.sp
Specifies the \fB\&.pem\fP file that contains the Certificate Revocation
List. Specify the file name of the \fB\&.pem\fP file using relative or
absolute paths.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslFIPSMode
New in version 2.6.

.sp
Directs the \fBbongo\fP to use the FIPS mode of the installed OpenSSL
library. Your system must have a FIPS compliant OpenSSL library to use
the \fI\-\-sslFIPSMode\fP option.
.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
FIPS Compatible SSL is
available only in \fI\%BongoDB Enterprise\fP\&. See
http://docs.bongodb.org/manual/tutorial/configure\-fips for more information.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslAllowInvalidCertificates
New in version 2.6.

.sp
Bypasses the validation checks for server certificates and allows
the use of invalid certificates. When using the
\fBallowInvalidCertificates\fP setting, BongoDB logs as a
warning the use of the invalid certificate.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslAllowInvalidHostnames
New in version 3.0.

.sp
Disables the validation of the hostnames in SSL certificates. Allows
\fBbongo\fP to connect to BongoDB instances if the hostname their
certificates do not match the specified hostname.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.SH FILES
.INDENT 0.0
.TP
.B \fB~/.dbshell\fP
\fBbongo\fP maintains a history of commands in the \fB\&.dbshell\fP
file.
.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
\fBbongo\fP does not recorded interaction related to
authentication in the history file, including
\fBauthenticate\fP and \fBdb.createUser()\fP\&.
.UNINDENT
.UNINDENT
.sp
\fBWARNING:\fP
.INDENT 7.0
.INDENT 3.5
Versions of Windows \fBbongo.exe\fP earlier than 2.2.0 will
save the \fI\&.dbshell\fP file in the \fBbongo.exe\fP working
directory.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \fB~/.bongorc.js\fP
\fBbongo\fP will read the \fB\&.bongorc.js\fP file from the home
directory of the user invoking \fBbongo\fP\&. In the file, users
can define variables, customize the \fBbongo\fP shell prompt,
or update information that they would like updated every time they
launch a shell. If you use the shell to evaluate a JavaScript file
or expression either on the command line with \fI\%\-\-eval\fP or
by specifying \fI\%a .js file to bongo\fP,
\fBbongo\fP will read the \fB\&.bongorc.js\fP file \fIafter\fP the
JavaScript has finished processing.
.sp
Specify the \fI\%\-\-norc\fP option to disable
reading \fB\&.bongorc.js\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \fB/etc/bongorc.js\fP
Global \fBbongorc.js\fP file which the \fBbongo\fP shell
evaluates upon start\-up. If a user also has a \fB\&.bongorc.js\fP
file located in the \fI\%HOME\fP directory, the \fBbongo\fP
shell evaluates the global \fB/etc/bongorc.js\fP file \fIbefore\fP
evaluating the user\(aqs \fB\&.bongorc.js\fP file.
.sp
\fB/etc/bongorc.js\fP must have read permission for the user
running the shell. The \fI\%\-\-norc\fP option for \fBbongo\fP
suppresses only the user\(aqs \fB\&.bongorc.js\fP file.
.sp
On Windows, the global \fBbongorc.js </etc/bongorc.js>\fP exists
in the \fB%ProgramData%\eBongoDB\fP directory.
.TP
.B \fB/tmp/bongo_edit\fI<time_t>\fP\&.js\fP
Created by \fBbongo\fP when editing a file. If the file exists,
\fBbongo\fP will append an integer from \fB1\fP to \fB10\fP to the
time value to attempt to create a unique file.
.TP
.B \fB%TEMP%bongo_edit\fI<time_t>\fP\&.js\fP
Created by \fBbongo.exe\fP on Windows when editing a file. If
the file exists, \fBbongo\fP will append an integer from \fB1\fP
to \fB10\fP to the time value to attempt to create a unique file.
.UNINDENT
.SH ENVIRONMENT
.INDENT 0.0
.TP
.B EDITOR
Specifies the path to an editor to use with the \fBedit\fP shell
command.  A JavaScript variable \fBEDITOR\fP will override the value of
\fI\%EDITOR\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B HOME
Specifies the path to the home directory where \fBbongo\fP will
read the \fB\&.bongorc.js\fP file and write the \fB\&.dbshell\fP

NEWS
WiredTiger release 2.9.0, 2016-09-06
------------------------------------

New features and API changes; refer to the API documentation for full details:

* WT-2360	Add nesting and disjunction functionality to cursor joins
* WT-2552	Add a public API to allow custom filesystem implementations
* WT-2738	Remove the ability to change the default checkpoint name. See the upgrading documentation for details.
* WT-2711	Change the statistics log configuration options. Restrict where statistics log files can be created to remove security vulnerability. See upgrading documentation for details.
* WT-2764	Enhance the checkpoint implementation, which led to a change in default eviction settings. See the upgrading documentation for details.
* WT-2880	Add Zstandard compression support
* See the upgrading documentation for information about renamed and removed statistics.

Significant changes and bug fixes:

* WT-2026	Fix a bug where in-memory pages were allowed to grow too large
* WT-2343	Ensure we cannot rename/drop when a backup cursor is open
* WT-2520	Ensure WT_SESSION::verify does not alter tables
* WT-2576	Ensure variable-length column-store cannot out-of-order return
* WT-2664	Change eviction so any eviction thread can find candidates
* WT-2702	Resolve an issue where under high thread load, WiredTiger exceeds cache size
* WT-2731	Resolve an issue where raw compression can create pages that are larger than expected
* WT-2737	Allow reconciliation to scrub dirty pages rather than evicting them
* WT-2757	Fix a bug in column stores where it was not possible to retrieve the new record number after an append
* WT-2816	Improve eviction performance

Issues fixed in MongoDB:

* SERVER-23504	Resolve a resource leak
* SERVER-23588	Stop using _open_osfhandle on Windows
* SERVER-23659	Improve log error message at startup
* SERVER-23661	Fix an issue where $sample takes disproportionately long time on newly created collection
* SERVER-24306	Fix stall in log_flush switching to new files
* SERVER-24580	Add more eviction stats to track efficiency
* SERVER-24580	Enhance eviction when application threads are contributing
* SERVER-24971	Don't cache buffers after application eviction
* SERVER-25843	Remove a redundant test
* SERVER-25845	Resolve an explicit null dereferenced
* SERVER-25846	Resolve a dereference after null with a check
* SERVER-26753	Don't spin on a read-lock in a tight loop

Other noteworthy changes since the previous release:

* WT-2103	Add incremental backup testing to test/format
* WT-2330	Ensure In-memory configurations do not create on-disk collection files
* WT-2450	Resolve an issue where salvage releases pages, then explicitly evicts them. Which can race with the eviction server
* WT-2453	Add multiple eviction queues
* WT-2504	Remove READONLY conditional for base config
* WT-2505	Resolve clang analyzer warnings
* WT-2508	Ensure test programs should remove test directories on the "clean" target
* WT-2512	Change how wtperf throttle values per thread are calculated
* WT-2513	Fix a type conversion warning
* WT-2517	Resolve incorrect uses of setvbuf on Windows
* WT-2518	Optimization of LSM checkpoint handle acquisition
* WT-2522	Resolve incorrect format code in message
* WT-2525	Miscellaneous cleanups of In-memory configurations
* WT-2526	Prevent mixing and matching readonly file-handles and read/write data-handles
* WT-2528	Fix a style error in WiredTiger build
* WT-2529	Be less aggressive asserting in readonly connections
* WT-2531	Resolve a case where in-memory tables are wasting space in truncation
* WT-2532	Change the WT_STREAM_APPEND and WT_STREAM_LINE_BUFFER flags as they overlap
* WT-2533	Don't let in-memory tables return a zero size
* WT-2534	Use atomic add when allocating transaction IDs
* WT-2535	Extend test/format to test for transactions reading their writes
* WT-2537	Check for an old and new uninitialized LSN for recovery
* WT-2539	Make streams a separate handle type from files
* WT-2540	Separate stream and file handle methods
* WT-2542	Resolve fixed-length column store reconciliation overwriting original values
* WT-2544	Fix eviction stats when clear is used
* WT-2546	Change evict server to be smarter when it decides whether or not to evict pages
* WT-2547	Add eviction testing configurations for 1 thread
* WT-2548	Cap the amount of data handed to raw compression
* WT-2550	Add support for raw mode cursors in java
* WT-2553	Add some In-memory documentation and stress testing
* WT-2556	Fix a typo in the Java example code
* WT-2557	Ensure that we truncate logs before closing backup cursor
* WT-2558	Reorganize WT_PAGE structure
* WT-2559	Resolve a potential segfault in Windows logging code
* WT-2560	Add a rwlock to protect transaction state, don't spin on this lock
* WT-2562	Add sleep loop to adapt for slow machines like PPC
* WT-2565	Further limit the amount of data passed to raw compression
* WT-2566	Add explicit memory barriers to Lock/unlock operations
* WT-2567	Add a check to see if logging is enabled for truncate rather than assuming it is
* WT-2568	Add a backward compatible constructor for PackInputStream
* WT-2569	Ensure __win_handle_read is followed by a call to GetLastError()
* WT-2570	Minor lint and cleanups to code
* WT-2571	Join code clean up
* WT-2572	Ensure test/format doesn't select an incompatible set of options for in-memory options configured
* WT-2573	Resolve a potential case where a double free can occur
* WT-2574	Ensure test/format frees all allocated configure memory
* WT-2576	Ensure variable-length column-store cannot out-of-order return
* WT-2577	Resolve a potential double-free following an in-memory split
* WT-2579	Change so diagnostic only code can use file streams directly
* WT-2580	Fix potential SWIG naming conflict in Java
* WT-2581	Resolve a case where we fail to track saved updates during a page split
* WT-2582	Fix a case where eviction can return WT_RESTART when running in-memory
* WT-2584	Change to not use periods at the end of error messages

src/bongo/shell/bongo.js
// bongo.js

// NOTE 'Bongo' may be defined here or in BongoJS.cpp.  Add code to init, not to this constructor.
if (typeof Bongo == "undefined") {
    Bongo = function(host) {
        this.init(host);
    };
}

if (!Bongo.prototype) {
    throw Error("Bongo.prototype not defined");
}

if (!Bongo.prototype.find)
    Bongo.prototype.find = function(ns, query, fields, limit, skip, batchSize, options) {
        throw Error("find not implemented");
    };
if (!Bongo.prototype.insert)
    Bongo.prototype.insert = function(ns, obj) {
        throw Error("insert not implemented");
    };
if (!Bongo.prototype.remove)
    Bongo.prototype.remove = function(ns, pattern) {
        throw Error("remove not implemented");
    };
if (!Bongo.prototype.update)
    Bongo.prototype.update = function(ns, query, obj, upsert) {
        throw Error("update not implemented");
    };

if (typeof bongoInject == "function") {
    bongoInject(Bongo.prototype);
}

Bongo.prototype.setSlaveOk = function(value) {
    if (value == undefined)
        value = true;
    this.slaveOk = value;
};

Bongo.prototype.getSlaveOk = function() {
    return this.slaveOk || false;
};

Bongo.prototype.getDB = function(name) {
    if ((jsTest.options().keyFile) &&
        ((typeof this.authenticated == 'undefined') || !this.authenticated)) {
        jsTest.authenticate(this);
    }
    // There is a weird issue where typeof(db._name) !== "string" when the db name
    // is created from objects returned from native C++ methods.
    // This hack ensures that the db._name is always a string.
    if (typeof(name) === "object") {
        name = name.toString();
    }
    return new DB(this, name);
};

Bongo.prototype.getDBs = function() {
    var res = this.adminCommand({"listDatabases": 1});
    if (!res.ok)
        throw _getErrorWithCode(res, "listDatabases failed:" + tojson(res));
    return res;
};

Bongo.prototype.adminCommand = function(cmd) {
    return this.getDB("admin").runCommand(cmd);
};

/**
 * Returns all log components and current verbosity values
 */
Bongo.prototype.getLogComponents = function() {
    var res = this.adminCommand({getParameter: 1, logComponentVerbosity: 1});
    if (!res.ok)
        throw _getErrorWithCode(res, "getLogComponents failed:" + tojson(res));
    return res.logComponentVerbosity;
};

/**
 * Accepts optional second argument "component",
 * string of form "storage.journaling"
 */
Bongo.prototype.setLogLevel = function(logLevel, component) {
    componentNames = [];
    if (typeof component === "string") {
        componentNames = component.split(".");
    } else if (component !== undefined) {
        throw Error("setLogLevel component must be a string:" + tojson(component));
    }
    var vDoc = {verbosity: logLevel};

    // nest vDoc
    for (var key, obj; componentNames.length > 0;) {
        obj = {};
        key = componentNames.pop();
        obj[key] = vDoc;
        vDoc = obj;
    }
    var res = this.adminCommand({setParameter: 1, logComponentVerbosity: vDoc});
    if (!res.ok)
        throw _getErrorWithCode(res, "setLogLevel failed:" + tojson(res));
    return res;
};

Bongo.prototype.getDBNames = function() {
    return this.getDBs().databases.map(function(z) {
        return z.name;
    });
};

Bongo.prototype.getCollection = function(ns) {
    var idx = ns.indexOf(".");
    if (idx < 0)
        throw Error("need . in ns");
    var db = ns.substring(0, idx);
    var c = ns.substring(idx + 1);
    return this.getDB(db).getCollection(c);
};

Bongo.prototype.toString = function() {
    return "connection to " + this.host;
};
Bongo.prototype.tojson = Bongo.prototype.toString;

/**
 * Sets the read preference.
 *
 * @param mode {string} read preference mode to use. Pass null to disable read
 *     preference.
 * @param tagSet {Array.<Object>} optional. The list of tags to use, order matters.
 *     Note that this object only keeps a shallow copy of this array.
 */
Bongo.prototype.setReadPref = function(mode, tagSet) {
    if ((this._readPrefMode === "primary") && (typeof(tagSet) !== "undefined") &&
        (Object.keys(tagSet).length > 0)) {
        // we allow empty arrays/objects or no tagSet for compatibility reasons
        throw Error("Can not supply tagSet with readPref mode primary");
    }
    this._setReadPrefUnsafe(mode, tagSet);
};

// Set readPref without validating. Exposed so we can test the server's readPref validation.
Bongo.prototype._setReadPrefUnsafe = function(mode, tagSet) {
    this._readPrefMode = mode;
    this._readPrefTagSet = tagSet;
};

Bongo.prototype.getReadPrefMode = function() {
    return this._readPrefMode;
};

Bongo.prototype.getReadPrefTagSet = function() {
    return this._readPrefTagSet;
};

// Returns a readPreference object of the type expected by bongos.
Bongo.prototype.getReadPref = function() {
    var obj = {}, mode, tagSet;
    if (typeof(mode = this.getReadPrefMode()) === "string") {
        obj.mode = mode;
    } else {
        return null;
    }
    // Server Selection Spec: - if readPref mode is "primary" then the tags field MUST
    // be absent. Ensured by setReadPref.
    if (Array.isArray(tagSet = this.getReadPrefTagSet())) {
        obj.tags = tagSet;
    }

    return obj;
};

/**
 * Sets the read concern.
 *
 * @param level {string} read concern level to use. Pass null to disable read concern.
 */
Bongo.prototype.setReadConcern = function(level) {
    if (!level) {
        this._readConcernLevel = undefined;
    } else if (level === "local" || level === "majority") {
        this._readConcernLevel = level;
    } else {
        throw Error("Invalid read concern.");
    }
};

/**
 * Gets the read concern.
 */
Bongo.prototype.getReadConcern = function() {
    return this._readConcernLevel;
};

connect = function(url, user, pass) {
    if (url instanceof BongoURI) {
        user = url.user;
        pass = url.password;
        url = url.uri;
    }
    if (user && !pass)
        throw Error("you specified a user and not a password.  " +
                    "either you need a password, or you're using the old connect api");

    // Validate connection string "url" as "hostName:portNumber/databaseName"
    //                                  or "hostName/databaseName"
    //                                  or "databaseName"
    //                                  or full bongo uri.
    var urlType = typeof url;
    if (urlType == "undefined") {
        throw Error("Missing connection string");
    }
    if (urlType != "string") {
        throw Error("Incorrect type \"" + urlType + "\" for connection string \"" + tojson(url) +
                    "\"");
    }
    url = url.trim();
    if (0 == url.length) {
        throw Error("Empty connection string");
    }

    if (!url.startsWith("bongodb://")) {
        const colon = url.lastIndexOf(":");
        const slash = url.lastIndexOf("/");
        if (slash == 0) {
            throw Error("Failed to parse bongodb:// URL: " + url);
        }
        if (slash == -1 && colon == -1) {
            url = "bongodb://127.0.0.1:27017/" + url;
        } else if (slash != -1) {
            url = "bongodb://" + url;
        }
    }

    chatty("connecting to: " + url);
    var m = new Bongo(url);
    db = m.getDB(m.defaultDB);

    if (user && pass) {
        if (!db.auth(user, pass)) {
            throw Error("couldn't login");
        }
    }

    // Check server version
    var serverVersion = db.version();
    chatty("BongoDB server version: " + serverVersion);

    var shellVersion = version();
    if (serverVersion.slice(0, 3) != shellVersion.slice(0, 3)) {
        chatty("WARNING: shell and server versions do not match");
    }

    return db;
};

/** deprecated, use writeMode below
 *
 */
Bongo.prototype.useWriteCommands = function() {
    return (this.writeMode() != "legacy");
};

Bongo.prototype.forceWriteMode = function(mode) {
    this._writeMode = mode;
};

Bongo.prototype.hasWriteCommands = function() {
    var hasWriteCommands = (this.getMinWireVersion() <= 2 && 2 <= this.getMaxWireVersion());
    return hasWriteCommands;
};

Bongo.prototype.hasExplainCommand = function() {
    var hasExplain = (this.getMinWireVersion() <= 3 && 3 <= this.getMaxWireVersion());
    return hasExplain;
};

/**
 * {String} Returns the current mode set. Will be commands/legacy/compatibility
 *
 * Sends isMaster to determine if the connection is capable of using bulk write operations, and
 * caches the result.
 */

Bongo.prototype.writeMode = function() {

    if ('_writeMode' in this) {
        return this._writeMode;
    }

    // get default from shell params
    if (_writeMode)
        this._writeMode = _writeMode();

    // can't use "commands" mode unless server version is good.
    if (this.hasWriteCommands()) {
        // good with whatever is already set
    } else if (this._writeMode == "commands") {
        this._writeMode = "compatibility";
    }

    return this._writeMode;
};

/**
 * Returns true if the shell is configured to use find/getMore commands rather than the C++ client.
 *
 * Currently, the C++ client will always use OP_QUERY find and OP_GET_MORE.
 */
Bongo.prototype.useReadCommands = function() {
    return (this.readMode() === "commands");
};

/**
 * For testing, forces the shell to use the readMode specified in 'mode'. Must be either "commands"
 * (use the find/getMore commands), "legacy" (use legacy OP_QUERY/OP_GET_MORE wire protocol reads),
 * or "compatibility" (auto-detect mode based on wire version).
 */
Bongo.prototype.forceReadMode = function(mode) {
    if (mode !== "commands" && mode !== "compatibility" && mode !== "legacy") {
        throw new Error("Mode must be one of {commands, compatibility, legacy}, but got: " + mode);
    }

    this._readMode = mode;
};

/**
 * Get the readMode string (either "commands" for find/getMore commands, "legacy" for OP_QUERY find
 * and OP_GET_MORE, or "compatibility" for detecting based on wire version).
 */
Bongo.prototype.readMode = function() {
    // Get the readMode from the shell params if we don't have one yet.
    if (typeof _readMode === "function" && !this.hasOwnProperty("_readMode")) {
        this._readMode = _readMode();
    }

    if (this.hasOwnProperty("_readMode") && this._readMode !== "compatibility") {
        // We already have determined our read mode. Just return it.
        return this._readMode;
    } else {
        // We're in compatibility mode. Determine whether the server supports the find/getMore
        // commands. If it does, use commands mode. If not, degrade to legacy mode.
        try {
            var hasReadCommands = (this.getMinWireVersion() <= 4 && 4 <= this.getMaxWireVersion());
            if (hasReadCommands) {
                this._readMode = "commands";
            } else {
                this._readMode = "legacy";
            }
        } catch (e) {
            // We failed trying to determine whether the remote node supports the find/getMore
            // commands. In this case, we keep _readMode as "compatibility" and the shell should
            // issue legacy reads. Next time around we will issue another isMaster to try to
            // determine the readMode decisively.
        }
    }

    return this._readMode;
};

//
// Write Concern can be set at the connection level, and is used for all write operations unless
// overridden at the collection level.
//

Bongo.prototype.setWriteConcern = function(wc) {
    if (wc instanceof WriteConcern) {
        this._writeConcern = wc;
    } else {
        this._writeConcern = new WriteConcern(wc);
    }
};

Bongo.prototype.getWriteConcern = function() {
    return this._writeConcern;
};

Bongo.prototype.unsetWriteConcern = function() {
    delete this._writeConcern;
};
